### DISCUSSION TITLE
regression & forecasting fine-tuning

### DISCUSSION THREAD

**Discussion by geoHeil** [timeseriesAI/tsai#133] [Q&A]
I have a dataset like this:
![grafik](https://user-images.githubusercontent.com/1694964/122361615-b6264d80-cf57-11eb-9434-1e284e33ad4d.png)

and for each multivariate panel time-series I want to forecast n steps ahead for the target variable baed on many features (the example here only shows a single one) as well as the history of the target variable itself. How do I need to define the `y_func`? 

So far I am simply using the sliding window panel fucntion with a specific forecast horizon
```

X,y = SlidingWindowPanel(window_len=216, 
                         unique_id_cols=['fnid'], 
                         stride=24, 
                         start=0, 
                         pad_remainder=False, 
                         padding_value=np.nan,
                         add_padding_feature=True,
                         get_x=['target', 'value_1'],
                         get_y='target',
                         y_func=None,
                         horizon=24,
                         seq_first=True,
                         sort_by=['time'],
                         ascending=True,
                         check_leakage=True,
                         return_key=False,
                         verbose=True
                        )(df)
```

Is this the correct way? Or do I need to perform any future considerations?
What I mean is: https://github.com/timeseriesAI/tsai/issues/36#issuecomment-752366817 I am unsure about the `get_y` if this is used correctly here as I want to use many dimensions of a multivariate time series to forecast a single dimension.

However, so far the results seem to be rather not well forecastable at all - the RMSE converges to 2.7 after a very short while of training.

**Comment by oguiza** [comment]
Hi,
Yes, as far as I can see, that'd be the correct way to do it (if you want to forecast the next 24 timesteps in the sequence). 
I'd like to suggest a way I often use to check whether my settings are correct. I normally create some dummy data that reflects the structure of the data I'm interested in and perform the operation to check the data is correct. I've created a simple [gist](https://gist.github.com/oguiza/1ac7b23e8a85a6052b39f0e6d817a0ed). 
BTW, I've also demonstrated a way to apply SlidingWindow on Panel data that I've found to be much faster that SlidingWindowPanel on some data. 
```
output = [df.groupby(['unique_id']).apply(lambda x: SlidingWindow(window_len=5,
                                                                  stride=5,
                                                                  start=0,
                                                                  pad_remainder=False,
                                                                  padding_value=np.nan,
                                                                  add_padding_feature=True,
                                                                  get_x=['var1', 'var2', 'target'],
                                                                  get_y='target',
                                                                  y_func=None,
                                                                  horizon=3,
                                                                  seq_first=True,
                                                                  sort_by=['time'],
                                                                  ascending=True,
                                                                  check_leakage=True,
                                                                  copy=True, # it's important to set copy to True when used in this way!!!
                                                                  )(x))][0].values

X = np.concatenate([oi[0] for oi in output])
y = np.concatenate([oi[1] for oi in output])
```

As to the training, there are multiple reasons why you are not getting the expected results. Are you scaling the data by variable using TSStandardize(by_var=True)? I've noticed the target and variable are in very different scales, which might have a large impact on the RMSE.

**Reply by meanpenguin** [reply]
Exactly what I needed...   
But....
As of tsai : 0.2.18, the copy argument is not supported in SlidingWindow()
https://github.com/timeseriesAI/tsai/blob/main/tsai/data/preparation.py#L308

**Reply by oguiza** [reply]
Hi @meanpenguin,
I'm sorry but I tested it in my local environment but forgot to push the commit to GitHub. SlidingWindow supports copy now.

**Reply by geoHeil** [reply]
I was playing with: TSStandardize(by_var=True, by_sample=True) and various combintions - but so far (including BERT pre-training) are not yet really happy with the results.

**Reply by geoHeil** [reply]
Great to hear that the overall setup should work. And thanks for the tip to speed up the panel data.

**Reply by geoHeil** [reply]
Indeed, the various variables are in very different ranges. However, I think to properly scale the data it is not that easy in my case: I have groups of panel devices and the scaling should be applied for each group.

**Reply by oguiza** [reply]
In that case, you may need to preprocess data before passing it to the data loader.

**Reply by geoHeil** [reply]
Normalizing per panel group certainly helps. But: 
![grafik](https://user-images.githubusercontent.com/1694964/123076234-38f84e00-d419-11eb-917a-2bd7e7e3687d.png)


it doe snot really solve the problem. Certainly the RMSE is now much better (0.6 vs. 2.7) but I still m not happy that the neural network basically stops learning after acouple of epochs (on the validation set) and is starting to overfit on training.

Do you have some suggestions of what I could try to improve the regression outcome?

**Reply by oguiza** [reply]
The model is clearly overfitting, so you can use any of the techniques to try and reduce it. Jeremy Howard recommends:
1. Use more data
2. Use data augmentation
3. More generalizable architecture
4. Regularization: dropout, weight decay
5. Reduce architecture complexity

The model seems to train well for the first 10-20 epochs. Then it starts overfitting. I'd set the number of epochs to 20 (max) to reduce training time and would use 1 (if you have more data) or 2 (you can pass CutMix1d() as a callback - it works well on many occasions).

In terms of architectures, I'm not sure which one you are using. I'd test at least MiniRocket, InceptionTime and TST.

You can also add dropout (passing fc_dropout=.5 for example to any of the *Plus models in `tsai`. 

Lastly, you can tune the model hyperparameters and reduce it's size.

**Reply by geoHeil** [reply]
When following your suggestions:

```
current_learner_model = create_model(InceptionTimePlus, dls=dls, **{'fc_dropout':0.5})
learn = Learner(dls, current_learner_model,  metrics=[mae, rmse], cbs=[ShowGraph(), CutMix1D()])
learn.fit_one_cycle(30, 1e-3)

```

the overfitting is certainly reduced. Notice:  I furthermore have reduced the forecasting horizion from 7 to 1 to make the task simpler and also switched to a randomized train-test split for the sliding windows.
I guess this is moving a little bit into the right direction. 
![grafik](https://user-images.githubusercontent.com/1694964/123271498-2fe2ac00-d501-11eb-91c2-1f6d09e82746.png)

**Reply by geoHeil** [reply]
Interestingly now after restarting the notebook I start to face: https://github.com/timeseriesAI/tsai/issues/136 and cannot find any reason for it.

**Reply by oguiza** [reply]
Are you sure the learner object was saved after the issue was fixed? That's strange. To delete the learner or recreate it shouldn't have any impact on the final learner object.

**Reply by meanpenguin** [reply]
Remember that the first time it works.  It’s the second time and on that it doesn’t.  You probably have not noticed it if you are not looking for it. :-).
